File: 2024/day14/day14.py
import re
def movie(positions, cols, rows, seconds):
    return [((c[0]+seconds*c[2])%cols, (c[1]+seconds*c[3])%rows) for c in positions]

def character(num):#to draw the tree
    if num>0:
        return '#'
    else:
        return '.'

with open("day14.txt", "r") as f:
    lines = f.read().split("\n")
pos_vel = [list(map(int, re.findall(r"\d+|-\d+", l))) for l in lines]

length = 101
height = 103
steps = 100

final_config = [((c[0]+steps*c[2])%length, (c[1]+steps*c[3])%height) for c in pos_vel]

first = [p for p in final_config if p[0]<(length-1)/2 and p[1]<(height-1)/2]
second = [p for p in final_config if p[0]>(length-1)/2 and p[1]<(height-1)/2]
third = [p for p in final_config if p[0]<(length-1)/2 and p[1]>(height-1)/2]
fourth = [p for p in final_config if p[0]>(length-1)/2 and p[1]>(height-1)/2]
print(len(first)*len(second)*len(third)*len(fourth))

to_check = movie(pos_vel, length, height, 1)
time_passed = 1571#starting from here because my input has two times where the robots occupy unique spaces
while len(set(to_check)) < len(to_check):
   time_passed += 1
   to_check = movie(pos_vel, length, height, time_passed)
grid = [[0 for _ in range(length)] for _ in range(height)]
for pos in to_check:
    grid[pos[1]][pos[0]] += 1

modified_grid = [[character(x) for x in r] for r in grid]
for r in modified_grid:
    print("".join(r))
print(time_passed)

File: 2024/day14/day14.hs
import Text.Regex.TDFA
import System.IO
import Data.List

movie :: Integral b => [[b]] -> b -> b -> b -> [(b, b)]
movie positions cols rows seconds = [(((c!!0)+seconds*(c!!2))`mod`cols, ((c!!1)+seconds*(c!!3))`mod`rows)|c<-positions]

partOne parsed_input cols rows =
    let
        final_config = movie parsed_input cols rows 100
        first = [p|p<-final_config, 2*(fst p)<(cols-1) && 2*(snd p)<(rows-1)]
        second = [p|p<-final_config, 2*(fst p)>(cols-1) && 2*(snd p)<(rows-1)]
        third = [p|p<-final_config, 2*(fst p)<(cols-1) && 2*(snd p)>(rows-1)]
        fourth = [p|p<-final_config, 2*(fst p)>(cols-1) && 2*(snd p)>(rows-1)]
    in
        (length first)*(length second)*(length third)*(length fourth)

partTwo parsed_input cols rows = [i|i<-[0..10000], length(nub(movie parsed_input cols rows i))== length(movie parsed_input cols rows i) ]

main = do
    handle <- openFile "day14.txt" ReadMode
    contents <- hGetContents handle
    let
        length = 101
        height = 103
        all = lines contents
        f string = (map (read::String->Int))(getAllTextMatches(string =~ "[0-9]+|-[0-9]+")::[String])
        pos_vel = [f l|l<-all]
    print(partOne pos_vel length height)
    print(partTwo pos_vel length height)